javascripts/2048.js

```diff
@@ -155,3 +155,117 @@ Board.prototype.moveBackward = function(condensedColOrRow) {
 Board.prototype.updateScore = function(points) {
   // this will somehow update the total score the player has going
 }
+
+//------------ UP example --------------
+
+// before reorient UP move
+board2 = new Board([
+  [2, 4,  32, 512],
+  [2, 4,  32, 256],
+  [0, 4, 128, 256],
+  [0, 2,  64,  32]
+]);
+// after reorient, before condense / resolve UP:
+//   [  2,   2,   0,   0]
+//   [  4,   4,   4,   2]
+//   [ 32,  32, 128,  64]
+//   [512, 256, 256,  32]
+
+// after condense & resolve -- now ready for build:
+var condensedUp = [
+  [4],
+  [8, 4, 2],
+  [64, 128, 64],
+  [512, 512, 32]
+];
+
+// after rebuild step
+// [  4,   0,  0, 0]
+// [  8,   4,  2, 0]
+// [ 64, 128, 64, 0]
+// [512, 512, 32, 0]
+
+// after reorient step, READY FOR NEW TILE! :)
+// [4, 8,  64, 512]
+// [0, 4, 128, 512]
+// [0, 2,  64,  32]
+// [0, 0,   0,   0]
+
+//------------ DOWN example --------------
+
+// before reorient DOWN move
+// [2, 4,  32, 512]
+// [2, 4,  32, 256]
+// [0, 4, 128, 256]
+// [0, 2,  64,  32]
+
+// after reorient, before condense / resolve DOWN:
+//   [  2,   2,   0,   0]
+//   [  4,   4,   4,   2]
+//   [ 32,  32, 128,  64]
+//   [512, 256, 256,  32]
+
+// after condense & resolve -- now ready for build:
+var condensedDown = [
+  [4],
+  [4, 8, 2],
+  [64, 128, 64],
+  [512, 512, 32]
+];
+
+// after rebuild step
+// [0,   0,   0,  4]
+// [0,   4,   8,  2]
+// [0,  64, 128, 64]
+// [0, 512, 512, 32]
+
+// after reorient step, READY FOR NEW TILE! :)
+// [0, 0,   0,   0]
+// [0, 4,  64, 512]
+// [0, 8, 128, 512]
+// [4, 2,  64,  32]
+
+Board.prototype.build = function(condensedArrays, direction) {
+  var emptySpots = []; // this will eventually be a set of [row, column] positions for all the 0s / empty spots
+  var boardLength = this.boardLength;
+  var emptyTile = this.emptyTile;
+
+  var rebuild = function(array, currentRow) { // currentRow is for emptySpots positions
+    while (array.length < boardLength) {
+      if (direction == "left" || direction == "up") {
+        // [2, 4] & we're about to push in this.emptyTile at index 2
+        currentColumn = array.length; // currentColumn is for emptySpots positions
+        array.push(emptyTile);
+      } else { // "right" || "down"
+        // [2, 4] & we're about to unshift in this.emptyTile at eventual index 1 (4 - 2 - 1 == 1)
+        currentColumn = boardLength - array.length - 1; // currentColumn is for emptySpots positions
+        array.unshift(emptyTile);
+      }; // see ya, if
+
+      emptySpots.push([currentRow, currentColumn]); // tell emptySpots where we just filled in an empty tile
+    } // see ya, while
+    return array;
+  } // see ya, rebuild()
```

Comment:
it has access to the outside scope, which we need in order to push the coordinates in to the `emptySpots` array.